src/cascade/model/model_writer.py

```diff
@@ -0,0 +1,356 @@
+                be within the min and max of the age and time tables, so this
+                passes in the extremal values.
+        """
+        self._session = session
+        self._dismod_file = session.dismod_file
+        if extremal_age_time:
+            self._ages = np.array(list(extremal_age_time[0]), dtype=np.float)
+            self._times = np.array(list(extremal_age_time[1]), dtype=np.float)
+        else:
+            self._ages = np.zeros((0,), dtype=np.float)
+            self._times = np.zeros((0,), dtype=np.float)
+        self._rate_rows = list()  # List of dictionaries for rates.
+        self._mulcov_rows = list()  # List of dictionaries for covariate multipliers.
+        self._rate_id = dict()  # The rate ids with the primary rates.
+        self._nslist = dict()  # rate to integer
+        self._nslist_pair_rows = list()  # list of nslist id, node, and smooth
+        self._covariate_id_func = lambda x: nan
+        self._flushed = False
+        self._children = None
+        self._clear_previous_model()
+
+    def _clear_previous_model(self):
+        """We could rewrite parts of tables that have changed. This is the
+        clear-cutting alternative."""
+        cleared_tables = list()
+        # Other tables are owned by the session.
+        owned_tables = ["age", "covariate", "mulcov", "nslist", "nslist_pair",
+                        "prior", "smooth_grid", "smooth", "time", "weight",
+                        "weight_grid"]
+        for create_name in owned_tables:
+            if not getattr(self._dismod_file, create_name).empty:
+                cleared_tables.append(create_name)
+                setattr(self._dismod_file, create_name, self._dismod_file.empty_table(create_name))
+
+        # Handle rate separately.
+        if not self._dismod_file.rate.empty:
+            cleared_tables.append("rate")
+            # Then the rate has all five entries.
+            self._dismod_file.rate.loc[:, "parent_smooth_id"] = nan
+            self._dismod_file.rate.loc[:, "child_smooth_id"] = nan
+            self._dismod_file.rate.loc[:, "child_nslist_id"] = nan
+        if cleared_tables:
+            CODELOG.debug(f"Writing model cleared tables from previous model: {', '.join(cleared_tables)}")
+
+    def start_model(self, nonzero_rates, children):
+        """To start the model, tell me enough to know the sizes almost all of
+        the tables by telling me the rate count and child random effect count."""
+        self._children = children
+        iota_case = "pos" if "iota" in nonzero_rates else "zero"
+        rho_case = "pos" if "rho" in nonzero_rates else "zero"
+        self._session.set_option(rate_case=f"iota_{iota_case}_rho_{rho_case}")
+
+    def write_ages_and_times(self, ages, times):
+        """
+        Collects all of the ages and times because they are ordered
+        and indexed by Dismod's file format. Can be called multiple times.
```

Comment:
Hmm, I see. This doesn't actually know the extremes because it may be called more than once. So you must either calculate the extremes outside of this or defer writing of all values that depend on age_id and time_id until you know this has been called as many times as it's going to be. Which means this whole system would need to be written as setup->accumulate->flush and that's a bit ugly. Hm...

Reply:
That's *exactly* what's happening.